File: 백준/Gold/2580. 스도쿠/스도쿠.py
import sys 
input = sys.stdin.readline
def rowSearch(row_num, sdoku_num):
    for c in range(9):
        if sdoku[row_num][c] == sdoku_num: #가로에 넣을 숫자가 이미 있음->False
            return False
    return True
def columnSearch(col_num, sdoku_num):
    for r in range(9):
        if sdoku[r][col_num] == sdoku_num: #세로에 넣을 숫자가 이미 있음->False
            return False
    return True
def squareSearch(row_num, col_num, sdoku_num): #사각형 안에 넣을 숫자가 이미 있음->False
    nr = (row_num//3) * 3
    nc = (col_num//3) * 3
    for i in range(3):
        for j in range(3):
            if sdoku[nr+i][nc+j] == sdoku_num:
                return False
    return True
def dfs(depth):
    if depth == len(zeros): #0끝까지 다 찾음
        for row in range(9):
            for col in range(9):
                print(sdoku[row][col], end=" ")
            print()
        exit()
    
    r, c = zeros[depth]	# 0의 위치
    for num in range(1, 10):
        if rowSearch(r, num) and columnSearch(c, num) and squareSearch(r, c, num):#세 곳 모두에 없음
            sdoku[r][c] = num
            dfs(depth + 1)
            sdoku[r][c] = 0

sdoku = []
zeros=[]
for i in range(9):
    tmp = list(map(int, input().split()))
    for j in range(9):
        if tmp[j]==0:
            zeros.append([i,j])
    sdoku.append(tmp)

dfs(0)

File: 백준/Gold/2580. 스도쿠/스도쿠.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

var target = make([][]int, 9)
var zeros = [][]int{}
var numbers = make([]int, 9)
var writer = bufio.NewWriter(os.Stdout)

func rowSearch(row, targetNum int) bool {
	for c := 0; c < 9; c++ {
		if target[row][c] == targetNum {
			return false
		}
	}
	return true
}

func colSearch(col, targetNum int) bool {
	for r := 0; r < 9; r++ {
		if target[r][col] == targetNum {
			return false
		}
	}
	return true
}

func squareSearch(row, col, targetNum int) bool {

	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			if target[row/3*3+i][col/3*3+j] == targetNum {
				return false
			}
		}
	}
	return true
}

func dfs(depth int) {
	if depth == len(zeros) {
		for i := 0; i < 9; i++ {
			for j := 0; j < 9; j++ {
				fmt.Fprintf(writer, "%d ", target[i][j])
			}
			fmt.Fprintf(writer, "\n")
		}
		writer.Flush()
		os.Exit(0)
	}
	v := zeros[depth]
	for num := 1; num <= 9; num++ {
		if rowSearch(v[0], num) && colSearch(v[1], num) && squareSearch(v[0], v[1], num) {
			target[v[0]][v[1]] = num
			dfs(depth + 1)
			target[v[0]][v[1]] = 0

		}
	}
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	target = make([][]int, 9)
	for i := 0; i < 9; i++ {
		target[i] = make([]int, 9)
		for j := 0; j < 9; j++ {
			fmt.Fscanf(reader, "%d ", &target[i][j])
			if target[i][j] == 0 {
				zeros = append(zeros, []int{i, j})
			}
		}
		numbers[i] = (i + 1)
	}
	dfs(0)
}
